Gardenlet panics when deployed to a managed seed

**How to categorize this issue?**

/area control-plane
/kind bug
/priority 3

**What happened**:

When deployed to a managed seed via a `ManagedSeed` resource, gardenlet panics:

```
panic: need an internal domain secret but found none

goroutine 1 [running]:
k8s.io/apimachinery/pkg/util/runtime.Must(...)
	/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:171
github.com/gardener/gardener/pkg/gardenlet/controller.(*GardenletControllerFactory).Run(0xc000f079b8, 0x256c860, 0xc0006f1280, 0xc000254300, 0xc00005e800)
	/go/src/github.com/gardener/gardener/pkg/gardenlet/controller/factory.go:145 +0x1ce5
github.com/gardener/gardener/cmd/gardenlet/app.(*Gardenlet).startControllers(0xc000991720, 0x256c860, 0xc0006f1280, 0x252d101, 0x251a740)
	/go/src/github.com/gardener/gardener/cmd/gardenlet/app/gardenlet.go:493 +0x152
github.com/gardener/gardener/cmd/gardenlet/app.(*Gardenlet).Run.func1(0x256c860, 0xc0006f1280, 0x256c860, 0xc0006f07c0)
	/go/src/github.com/gardener/gardener/cmd/gardenlet/app/gardenlet.go:445 +0x65
github.com/gardener/gardener/cmd/gardenlet/app.(*Gardenlet).Run(0xc000991720, 0x256c860, 0xc00052ecc0, 0x0, 0x0)
	/go/src/github.com/gardener/gardener/cmd/gardenlet/app/gardenlet.go:475 +0x790
github.com/gardener/gardener/cmd/gardenlet/app.run(0x256c860, 0xc00052ecc0, 0xc0003c08a0, 0x1db7840, 0xc0007d5d60)
	/go/src/github.com/gardener/gardener/cmd/gardenlet/app/gardenlet.go:171 +0x10a
github.com/gardener/gardener/cmd/gardenlet/app.NewCommandStartGardenlet.func1(0xc0007ec840, 0xc00028e6a0, 0x0, 0x1, 0x0, 0x0)
	/go/src/github.com/gardener/gardener/cmd/gardenlet/app/gardenlet.go:201 +0x99
github.com/spf13/cobra.(*Command).execute(0xc0007ec840, 0xc00004c070, 0x1, 0x1, 0xc0007ec840, 0xc00004c070)
	/go/src/github.com/gardener/gardener/vendor/github.com/spf13/cobra/command.go:850 +0x472
github.com/spf13/cobra.(*Command).ExecuteC(0xc0007ec840, 0xc00052ecc0, 0x0, 0x1)
	/go/src/github.com/gardener/gardener/vendor/github.com/spf13/cobra/command.go:958 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/go/src/github.com/gardener/gardener/vendor/github.com/spf13/cobra/command.go:895
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	/go/src/github.com/gardener/gardener/vendor/github.com/spf13/cobra/command.go:888
main.main()
	/go/src/github.com/gardener/gardener/cmd/gardenlet/main.go:36 +0x12c
```

**What you expected to happen**:
No panic.

**How to reproduce it (as minimally and precisely as possible)**:
Create a `ManagedSeed` resource with `gardenlet`, or use the `use-as-seed` annotation.

**Anything else we need to know?**:

**Environment**:

- Gardener version: master (d6914db40e716cb1e60c5e726f64ffea4405d8ee)
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

--- a/pkg/gardenlet/controller/factory.go
+++ b/pkg/gardenlet/controller/factory.go
@@ -19,6 +19,7 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
+	"time"
 
 	"github.com/gardener/gardener/charts"
 	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
@@ -42,10 +43,13 @@ import (
 	"github.com/gardener/gardener/pkg/logger"
 	"github.com/gardener/gardener/pkg/operation/garden"
 	seedpkg "github.com/gardener/gardener/pkg/operation/seed"
+	"github.com/gardener/gardener/pkg/utils"
 	"github.com/gardener/gardener/pkg/utils/imagevector"
 	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	retryutils "github.com/gardener/gardener/pkg/utils/retry"
 
 	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/util/runtime"
@@ -53,6 +57,8 @@ import (
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/tools/record"
 	"k8s.io/component-base/version"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 )
 
 // DefaultImageVector is a constant for the path to the default image vector file.
@@ -134,8 +140,16 @@ func (f *GardenletControllerFactory) Run(ctx context.Context) error {
 		return fmt.Errorf("no seed selected by Gardenlet")
 	}
 
+	// Register Seed object if desired
+	if f.cfg.SeedConfig != nil {
+		if err := f.registerSeed(ctx, k8sGardenClient.Client()); err != nil {
+			return fmt.Errorf("failed to register the seed: %+v", err)
+		}
+	}
+
 	// Read Garden secrets from any seed namespace as we assume they are synced accordingly by the Gardener Controller Manager.
 	// This requires adaption if we'll decide not to sync all Secrets for all Seeds in the future.
+	// For newly created seed, this seed namespace is about to be created by the GCM.
 	_, err = garden.ReadGardenSecrets(
 		ctx,
 		k8sGardenClient.Cache(),
@@ -239,4 +253,48 @@ func seedNames(seedConfig *config.SeedConfig, seedLister gardencorelisters.SeedL
 	}
 
 	return names
+}
+
+// registerSeed create or update the seed resource if gardenlet is configured to takes care about it.
+func (f *GardenletControllerFactory) registerSeed(ctx context.Context, gardenClient client.Client) error {
+	var (
+		seed = &gardencorev1beta1.Seed{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: f.cfg.SeedConfig.Name,
+			},
+		}
+		seedNamespaceName = seedpkg.ComputeGardenNamespace(f.cfg.SeedConfig.Name)
+		seedNamespace     = &corev1.Namespace{}
+	)
+
+	// Convert gardenlet config to an external version
+	cfg, err := confighelper.ConvertGardenletConfigurationExternal(f.cfg)
+	if err != nil {
+		return fmt.Errorf("could not convert gardenlet configuration: %+v", err)
+	}
+
+	if _, err := controllerutil.CreateOrUpdate(ctx, gardenClient, seed, func() error {
+		seed.Labels = utils.MergeStringMaps(map[string]string{
+			v1beta1constants.GardenRole: v1beta1constants.GardenRoleSeed,
+		}, f.cfg.SeedConfig.Labels)
+
+		seed.Spec = cfg.SeedConfig.Spec
+		return nil
+	}); err != nil {
+		return fmt.Errorf("could not register seed %q: %+v", seed.Name, err)
+	}
+
+	// wait seed namespace to be created by GCM
+	return retryutils.UntilTimeout(ctx, 5*time.Second, 2*time.Minute, func(context.Context) (done bool, err error) {
+		if err := gardenClient.Get(ctx, kutil.Key(seedNamespaceName), seedNamespace); err != nil {
+			if apierrors.IsNotFound(err) {
+				logger.Logger.Infof("Waiting until namespace %q is created.", seedNamespaceName)
+				return retryutils.MinorError(fmt.Errorf("namespace %q still not created", seedNamespaceName))
+			}
+			return retryutils.SevereError(err)
+		}
+
+		logger.Logger.Infof("Namespace %q has been created.", seedNamespaceName)
+		return retryutils.Ok()
+	})
 }

--- a/pkg/gardenlet/controller/seed/seed.go
+++ b/pkg/gardenlet/controller/seed/seed.go
@@ -16,7 +16,6 @@ package seed
 
 import (
 	"context"
-	"fmt"
 	"sync"
 	"time"
 
@@ -28,14 +27,11 @@ import (
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/tools/record"
 	"k8s.io/client-go/util/workqueue"
-	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 
 	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
-	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
 	gardencoreinformers "github.com/gardener/gardener/pkg/client/core/informers/externalversions"
 	gardencorelisters "github.com/gardener/gardener/pkg/client/core/listers/core/v1beta1"
 	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap"
-	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap/keys"
 	"github.com/gardener/gardener/pkg/controllerutils"
 	"github.com/gardener/gardener/pkg/gardenlet"
 	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
@@ -43,7 +39,6 @@ import (
 	"github.com/gardener/gardener/pkg/gardenlet/controller/lease"
 	"github.com/gardener/gardener/pkg/healthz"
 	"github.com/gardener/gardener/pkg/logger"
-	"github.com/gardener/gardener/pkg/utils"
 	"github.com/gardener/gardener/pkg/utils/imagevector"
 )
 
@@ -173,33 +168,6 @@ func (c *Controller) Run(ctx context.Context, workers int) {
 	}()
 
 	logger.Logger.Info("Seed controller initialized.")
-
-	// Register Seed object if desired
-	if c.config.SeedConfig != nil {
-		// Convert gardenlet config to an external version
-		cfg, err := confighelper.ConvertGardenletConfigurationExternal(c.config)
-		if err != nil {
-			panic(fmt.Errorf("could not convert gardenlet configuration: %+v", err))
-		}
-
-		seed := &gardencorev1beta1.Seed{ObjectMeta: metav1.ObjectMeta{Name: c.config.SeedConfig.Name}}
-
-		gardenClient, err := c.clientMap.GetClient(ctx, keys.ForGarden())
-		if err != nil {
-			panic(fmt.Errorf("could not register seed %q: failed to get garden client: %+v", seed.Name, err))
-		}
-
-		if _, err := controllerutil.CreateOrUpdate(ctx, gardenClient.Client(), seed, func() error {
-			seed.Labels = utils.MergeStringMaps(map[string]string{
-				v1beta1constants.GardenRole: v1beta1constants.GardenRoleSeed,
-			}, c.config.SeedConfig.Labels)
-
-			seed.Spec = cfg.SeedConfig.Spec
-			return nil
-		}); err != nil {
-			panic(fmt.Errorf("could not register seed %q: %+v", seed.Name, err))
-		}
-	}
 
 	for i := 0; i < workers; i++ {
 		controllerutils.DeprecatedCreateWorker(ctx, c.seedQueue, "Seed", c.reconcileSeedKey, &waitGroup, c.workerCh)

